### PR TITLE
fix(backend): Add membersCount to Organization

### PR DESCRIPTION
**Pull request by desiprisg**
## Type of change

- [x] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [x] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

Add missing properties to the Organization in clerk/backend. Avoid the privateMetadata property as it is not available on the client (from FAPI).

### CODE REVIEW

File: packages/backend/src/api/resources/JSON.ts [hunk 1]
```diff
@@ -127,6 +127,8 @@ export interface OrganizationJSON extends ClerkResourceJSON {
   name: string;
   slug: string | null;
   logo_url: string | null;
+  membersCount: number;
+  pendingInvitationsCount: number;
```
Comment (gkats) [inline packages/backend/src/api/resources/JSON.ts:131]: :wrench: The JSON response will return `pending_invitations_count` (snake cased) in some cases. Although this is never returned from BAPI as far as I know.

**Review by gkats — CHANGES_REQUESTED** [review]
Where are these values used? Are we certain we want to add them to the `Organization` resource that comes from BAPI?

File: packages/backend/src/api/resources/Organization.ts [hunk 2]
```diff
@@ -6,6 +6,8 @@ export class Organization {
     readonly name: string,
     readonly slug: string | null,
     readonly logoUrl: string | null,
+    readonly membersCount: number,
+    readonly pendingInvitationsCount: number,
```
[the remark this answers is not in the document]
Comment (gkats) [reply packages/backend/src/api/resources/Organization.ts:10]: The attribute might also be present when an organization object is nested inside an organization membership object in BAPI. I can double check unless we're certain we want it here. In that case, let's just handle the fact that the value might not be present in every response.

**Comment by desiprisg** [comment]
Thanks for the comments @gkats. I made this PR exactly for this reason, as we want to align the OrganizationType from clerk/backend and the types package. Does it sound feasible?

**Comment by gkats** [comment]
> Does it sound feasible?

@desiprisg It certainly does sound feasible, thanks for explaining. Let's just be careful as these two properties are usually not present in most organization related responses.

**Review by dimkl — APPROVED** [review]
🔧  i think we should also add it to : https://github.com/clerkinc/javascript/blob/c11217cf4432aa640af67ab2fd48f4e5bb33ce6d/packages/types/src/json.ts#L257

**Comment by desiprisg** [comment]
@dimkl Which one? Do you mean the null value for `members_count`?

**Comment by dimkl** [comment]
> @dimkl Which one? Do you mean the null value for `members_count`?

We should define the same as `packages/backend/src/api/resources/JSON.ts`

**Comment by desiprisg** [comment]
> > @dimkl Which one? Do you mean the null value for `members_count`?
> 
> We should define the same as `packages/backend/src/api/resources/JSON.ts`

It's coming from FAPI though. And they're not always the same, right? Do we just make them optional?

**Comment by desiprisg** [comment]
This isn't an "alignment" any more. We just added a missing field for now. I think we should discuss how we want to approach this further.

**Review by gkats — APPROVED** [review]

**Comment by dimkl** [comment]
@desiprisg I think that's both in FAPI and in BAPI in some endpoints of organisation (where organization memberships are loaded or organization is updated/created).
@gkats could you confirm if BAPI response is always a superset of related FAPI response?

**Review by nikosdouvlis — APPROVED** [review]

**Comment by gkats** [comment]
> could you confirm if BAPI response is always a superset of related FAPI response?

Not necessarily. 

Even if it is, I believe we should never rely on this. FAPI and BAPI are allowed to have completely different responses for the same resource.

This is by design. The name of the resource might be the same, but in a different context each resource should be independent.

IMO inheritance would be the wrong approach here. Distinct types for FAPI and BAPI resources is the way to go. Any duplication is healthy.

cc @SokratisVidros and @nikosdouvlis Since you have more knowledge on our FE libs and types package, what do you think on the above?

**Comment by nikosdouvlis** [comment]
> > could you confirm if BAPI response is always a superset of related FAPI response?
> 
> Not necessarily.
> 
> Even if it is, I believe we should never rely on this. FAPI and BAPI are allowed to have completely different responses for the same resource.
> 
> This is by design. The name of the resource might be the same, but in a different context each resource should be independent.
> 
> IMO inheritance would be the wrong approach here. Distinct types for FAPI and BAPI resources is the way to go. Any duplication is healthy.
> 
> cc @SokratisVidros and @nikosdouvlis Since you have more knowledge on our FE libs and types package, what do you think on the above?

I agree. Even though these types are very similar, I think decoupling the FE/BE types by introducing a little duplication makes sense. The differences are small, but they are quite important as they deal with security/privacy concerns (eg `privateMetadata`).

Another reason I'd keep them decoupled for now is that with the whole "move back to the server" (Next13 /app, React Server Components, SSR etc) movement, its hard to predict what architectural changes we'll need to do in the following months. The differentiation between BE and FE is a bit blurry right now, as our code needs to run in many different contexts (server, client, client hydration with server data, etc) that deal with different kind of resources (FAPI vs BAPI). For example, a RSC can access a BAPI resource while rendering on the server, but once it gets sent to the browser, it needs to be limited to the corresponding FAPI resource. This RSC, however, would be defined in a single file -  so we'll need to figure a way to make our types makes sense in this new ecosystem.

**Comment by dimkl** [comment]
By checking the response of the following BAPI endpoint we see that the `members_count` is not included. We can proceed with merging this PR without making a change to `backend` package.
```
curl 'https://api.clerk.dev/v1/organizations' \
    -H "Authorization: Bearer $SECRET_KEY" \
    --compressed
```

**Comment by desiprisg** [comment]
Even though we state in the docs that it is included, @dimkl is right. I'm closing this without merging.

**Pull request closed without merging**